### PR TITLE
Duet AI Prompts - Documentation Lookup Without Links 

### PR DESCRIPTION
**Pull request by dariabezkorovaina**
* Creating versions without links for the existing documentation lookup prompts
* Nits and fixes to documentation lookup prompts to improve language and match Google Writing Style guide 
 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by dariabezkorovaina** [comment]
@andreydevyatkin please review when you have a chance.

File: learning/prompts/documentation-lookup-nolinks/18_io_kafka.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+Here is an example of Apache Beam pipeline code for reading data from a Kafka topic:
+
+```python
+with beam.Pipeline(options=options) as p:
+    output = (p | "Read from Kafka" >> ReadFromKafka(
+                        consumer_config={'bootstrap.servers': 'localhost:9092'},
+                        topics=['my_topic'],
+                        with_metadata=False)
+                        | "Log Data" >> Map(logging.info))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/18_io_kafka.md:19]:
```suggestion
with beam.Pipeline(options=options) as p:
    output = (
        p
        | "Read from Kafka"
        >> ReadFromKafka(
            consumer_config={"bootstrap.servers": "localhost:9092"},
            topics=["my_topic"],
            with_metadata=False,
        )
        | "Log Data" >> Map(logging.info)
    )
```

File: learning/prompts/documentation-lookup-nolinks/08_basic_windowing.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+Because Apache Beam's default windowing strategy is to assign each element to a single, global window, you must explicitly specify a windowing function for your pipeline.
+
+The following code snippet shows how to divide a `PCollection` into 60-second windows:
+
+```python
+from apache_beam import beam
+from apache_beam import window
+fixed_windowed_items = (
+    items | 'window' >> beam.WindowInto(window.FixedWindows(60)))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/08_basic_windowing.md:17]:
```suggestion
from apache_beam import beam
from apache_beam import window

fixed_windowed_items = items | 'window' >> beam.WindowInto(window.FixedWindows(60))
```

File: learning/prompts/documentation-lookup-nolinks/19_io_pubsub.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+Here is an example of the Apache Beam pipeline code for reading data from a Pub/Sub topic and writing it to another topic:
+
+```python
+from apache_beam.io import ReadFromPubSub
+from apache_beam.io import WriteToPubSub
+
+  with beam.Pipeline(options=options) as p:
+    (p | "Read from Pub/Sub" >> ReadFromPubSub(topic="input_topic")
+       | "Write to Pub/Sub" >> WriteToPubSub(topic="output_topic"))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/19_io_pubsub.md:19]:
```suggestion
from apache_beam.io import ReadFromPubSub
from apache_beam.io import WriteToPubSub

with beam.Pipeline(options=options) as p:
    (
        p
        | "Read from Pub/Sub" >> ReadFromPubSub(topic="input_topic")
        | "Write to Pub/Sub" >> WriteToPubSub(topic="output_topic")
    )
```

File: learning/prompts/documentation-lookup-nolinks/20_io_biguery.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+Here is an example of Apache Beam pipeline code for reading from BigQuery:
+
+```python
+from apache_beam.io.gcp.bigquery import ReadFromBigQuery
+
+with beam.Pipeline(options=options) as p:
+  # read from a table
+    lines_table = p | 'Read' >> ReadFromBigQuery(table=table)
+  # read from a query
+    lines_query = p | 'Read' >> ReadFromBigQuery(query="SELECT * FROM table")
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/20_io_biguery.md:24]:
```suggestion
from apache_beam.io.gcp.bigquery import ReadFromBigQuery

with beam.Pipeline(options=options) as p:
    # read from a table
    lines_table = p | "Read" >> ReadFromBigQuery(table=table)
    # read from a query
    lines_query = p | "Read" >> ReadFromBigQuery(query="SELECT * FROM table")
```

File: learning/prompts/documentation-lookup-nolinks/20_io_biguery.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+```python
+from apache_beam.io.gcp.bigquery import WriteToBigQuery
+
+with beam.Pipeline(options=options) as p:
+  # write to a table
+    p | 'Write' >> beam.io.WriteToBigQuery(
+        table,
+        schema=TABLE_SCHEMA,
+        create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
+        write_disposition=beam.io.BigQueryDisposition.WRITE_APPEND)
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/20_io_biguery.md:38]:
```suggestion
from apache_beam.io.gcp.bigquery import WriteToBigQuery

with beam.Pipeline(options=options) as p:
    # write to a table
    p | "Write" >> beam.io.WriteToBigQuery(
        table,
        schema=TABLE_SCHEMA,
        create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
        write_disposition=beam.io.BigQueryDisposition.WRITE_APPEND,
    )
```

File: learning/prompts/documentation-lookup-nolinks/21_io_bigtable.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+```python
+with beam.Pipeline(options=options) as p:
+output = (p | "Read from Bigtable" >> ReadFromBigtable(
+                    project_id=project_id,
+                    instance_id=instance_id,
+                    table_id=table_id,
+                 )
+                    | "Extract cells" >> beam.Map(lambda row: row._cells)
+                    | "Log Data" >> Map(logging.info))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/21_io_bigtable.md:21]:
```suggestion
with beam.Pipeline(options=options) as p:
    output = (
        p
        | "Read from Bigtable"
        >> ReadFromBigtable(
            project_id=project_id,
            instance_id=instance_id,
            table_id=table_id,
        )
        | "Extract cells" >> beam.Map(lambda row: row._cells)
        | "Log Data" >> Map(logging.info)
    )
```

File: learning/prompts/documentation-lookup-nolinks/22_io_spanner.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+with beam.Pipeline(options=options) as p:
+    output = (p | "Read from table" >> ReadFromSpanner(
+        project_id=options.project_id,
+        instance_id=options.instance_id,
+        database_id=options.database_id,
+        row_type=ExampleRow,
+        sql="SELECT * FROM example_row"
+        )
+        | "Map Data" >> Map(lambda row: f"Id = {row.id}, Name = {row.name}")
+        | "Log Data" >> Map(logging.info))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/22_io_spanner.md:27]:
```suggestion
class ExampleRow(NamedTuple):
    id: int
    name: str


with beam.Pipeline(options=options) as p:
    output = (
        p
        | "Read from table"
        >> ReadFromSpanner(
            project_id=options.project_id,
            instance_id=options.instance_id,
            database_id=options.database_id,
            row_type=ExampleRow,
            sql="SELECT * FROM example_row",
        )
        | "Map Data" >> Map(lambda row: f"Id = {row.id}, Name = {row.name}")
        | "Log Data" >> Map(logging.info)
    )
```

File: learning/prompts/documentation-lookup-nolinks/24_io_json.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            help='Json file path'
+        )
+
+options = JsonOptions()
+with beam.Pipeline(options=options) as p:
+    output = (p | "Read from Json file" >> ReadFromJson(
+        path=options.file_path,
+        lines=False
+        )
+        | "Log Data" >> Map(logging.info))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/24_io_json.md:27]:
```suggestion
class JsonOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        parser.add_argument(
            "--file_path",
            default="gs://your-bucket/your-file.json",
            help="Json file path",
        )


options = JsonOptions()
with beam.Pipeline(options=options) as p:
    output = (
        p
        | "Read from Json file" >> ReadFromJson(path=options.file_path, lines=False)
        | "Log Data" >> Map(logging.info)
    )
```

File: learning/prompts/documentation-lookup-nolinks/25_io_csv.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        )
+
+options = CsvOptions()
+
+with beam.Pipeline(options=options) as p:
+    output = (p | "Read from Csv file" >> ReadFromCsv(
+        path=options.file_path
+        )
+        | "Log Data" >> Map(logging.info))
+
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/25_io_csv.md:28]:
```suggestion
class CsvOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        parser.add_argument(
            "--file_path",
            default="gs://your-bucket/your-file.csv",
            help="Csv file path",
        )


options = CsvOptions()

with beam.Pipeline(options=options) as p:
    output = (
        p
        | "Read from Csv file" >> ReadFromCsv(path=options.file_path)
        | "Log Data" >> Map(logging.info)
    )
```

File: learning/prompts/documentation-lookup-nolinks/26_io_avro.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+          "--path",
+          default="gs://cloud-samples-data/bigquery/us-states/*.avro",
+          help="GCS path to read from")
+
+  options = ReadAvroOptions()
+
+  with beam.Pipeline(options=options) as p:
+
+    (p | "Read from Avro" >> ReadFromAvro(options.path)
+       | Map(logging.info))
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/26_io_avro.md:26]:
```suggestion
class ReadAvroOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        parser.add_argument(
            "--path",
            default="gs://cloud-samples-data/bigquery/us-states/*.avro",
            help="GCS path to read from",
        )

    options = ReadAvroOptions()

    with beam.Pipeline(options=options) as p:
        (p | "Read from Avro" >> ReadFromAvro(options.path) | Map(logging.info))
```

File: learning/prompts/documentation-lookup-nolinks/27_io_parquet.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+          "--path",
+          help="GCS path to parquet file")
+
+  options = ReadParquetOptions()
+
+  with beam.Pipeline(options=options) as p:
+
+    (p | "ReadParquet" >> ReadFromParquet(options.path)
+       | Map(logging.info))
+
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/27_io_parquet.md:26]:
```suggestion
class ReadParquetOptions(PipelineOptions):
    @classmethod
    def _add_argparse_args(cls, parser):
        parser.add_argument("--path", help="GCS path to parquet file")

    options = ReadParquetOptions()

    with beam.Pipeline(options=options) as p:
        (p | "ReadParquet" >> ReadFromParquet(options.path) | Map(logging.info))
```

File: learning/prompts/documentation-lookup-nolinks/29_advanced_side_inputs.md [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+PCollection<Integer> input = ...;
+PCollectionView<Integer> sideInput = ...;
+PCollection<Integer> output =
+    input.apply(ParDo.of(new DoFn<Integer, Integer>() {
+      @ProcessElement
+      public void processElement(ProcessContext c) {
+        Integer sideInputValue = c.sideInput(sideInput);
+        ...
+      }
+    }).withSideInputs(sideInput));
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/29_advanced_side_inputs.md:29]:
```suggestion
PCollection<Integer> input = ...;
PCollectionView<Integer> sideInput = ...;
PCollection<Integer> output =
    input.apply(ParDo.of(new DoFn<Integer, Integer>() {
        @ProcessElement
        public void processElement(ProcessContext c) {
            Integer sideInputValue = c.sideInput(sideInput);
            ...
        }
    }).withSideInputs(sideInput));
```

File: learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+         out.get(mainOutputTag).output(word);
+       } else {
+         // Emit element to additional string output
+         out.get(additionalOutputTagString).output(word);
+       }
+       if (condition for additional integer output) {
+         // Emit element to additional integer output
+         out.get(additionalOutputTagIntegers).output(word.length());
+       }
+     }
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md:52]:
```suggestion
public void processElement(@Element String word, MultiOutputReceiver out) {
    if (condition
        for main output) {
        // Emit element to main output
        out.get(mainOutputTag).output(word);
    } else {
        // Emit element to additional string output
        out.get(additionalOutputTagString).output(word);
    }
    if (condition
        for additional integer output) {
        // Emit element to additional integer output
        out.get(additionalOutputTagIntegers).output(word.length());
    }
}
```

File: learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    yield pvalue.TaggedOutput(self.OUTPUT_TAG_CHARACTER_COUNT, len(element))
+
+    words = re.findall(r'[A-Za-z\']+', element)
+    for word in words:
+      if len(word) <= 3:
+        # yield word as an output to the OUTPUT_TAG_SHORT_WORDS tagged collection.
+        yield pvalue.TaggedOutput(self.OUTPUT_TAG_SHORT_WORDS, word)
+      else:
+        # yield word to add it to the main collection.
+        yield word
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md:77]:
```suggestion
class SplitLinesToWordsFn(beam.DoFn):
    # These tags will be used to tag the outputs of this DoFn.
    OUTPUT_TAG_SHORT_WORDS = "tag_short_words"
    OUTPUT_TAG_CHARACTER_COUNT = "tag_character_count"

    def process(self, element):
        # yield a count (integer) to the OUTPUT_TAG_CHARACTER_COUNT tagged collection.
        yield pvalue.TaggedOutput(self.OUTPUT_TAG_CHARACTER_COUNT, len(element))

        words = re.findall(r"[A-Za-z\']+", element)
        for word in words:
            if len(word) <= 3:
                # yield word as an output to the OUTPUT_TAG_SHORT_WORDS tagged collection.
                yield pvalue.TaggedOutput(self.OUTPUT_TAG_SHORT_WORDS, word)
            else:
                # yield word to add it to the main collection.
                yield word
```

File: learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        lines
+        | beam.ParDo(SplitLinesToWordsFn()).with_outputs(
+            SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS,
+            SplitLinesToWordsFn.OUTPUT_TAG_CHARACTER_COUNT,
+            main='words'))
+
+    # split_lines_result is an object of type DoOutputsTuple
+    words, _, _ = split_lines_result
+    short_words = split_lines_result[SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS]
+    character_count = split_lines_result.tag_character_count
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md:98]:
```suggestion
with beam.Pipeline(options=pipeline_options) as p:
    lines = p | ReadFromText(known_args.input)

    # with_outputs allows accessing the explicitly tagged outputs of a DoFn.
    split_lines_result = lines | beam.ParDo(SplitLinesToWordsFn()).with_outputs(
        SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS,
        SplitLinesToWordsFn.OUTPUT_TAG_CHARACTER_COUNT,
        main="words",
    )

    # split_lines_result is an object of type DoOutputsTuple
    words, _, _ = split_lines_result
    short_words = split_lines_result[SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS]
    character_count = split_lines_result.tag_character_count
```

File: learning/prompts/documentation-lookup-nolinks/53_dataflow_runner.md [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+For the Apache Beam Java SDK, in your terminal, run the following command (from your `word-count-beam` directory):
+
+```java
+  mvn -Pdataflow-runner compile exec:java \
+    -Dexec.mainClass=org.apache.beam.examples.WordCount \
+    -Dexec.args="--project=PROJECT_ID \
+    --gcpTempLocation=gs://BUCKET_NAME/temp/ \
+    --output=gs://BUCKET_NAME/output \
+    --runner=DataflowRunner \
+    --region=REGION
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/53_dataflow_runner.md:57]:
```suggestion
mvn -Pdataflow-runner compile exec:java \
  -Dexec.mainClass=org.apache.beam.examples.WordCount \
  -Dexec.args="--project=PROJECT_ID \
  --gcpTempLocation=gs://BUCKET_NAME/temp/ \
  --output=gs://BUCKET_NAME/output \
  --runner=DataflowRunner \
  --region=REGION
```

File: learning/prompts/documentation-lookup-nolinks/53_dataflow_runner.md [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+   ```
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/53_dataflow_runner.md:58]: nit: remove extra spaces

File: learning/prompts/documentation-lookup-nolinks/54_spark_runner.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+For additional pipeline configuration options, refer to the Spark Runner documentation.
+
+***3. Run Your Pipeline:***
+
+In Python, you can use the `run()` method of your pipeline object to execute the pipeline:
+
+```python
+    # Run your pipeline
+    p.run()
```
Comment (andreydevyatkin) [inline learning/prompts/documentation-lookup-nolinks/54_spark_runner.md:141]:
```suggestion
# Run your pipeline
p.run()
```

**Review by andreydevyatkin — COMMENTED** [review]
See the inline comments

**Comment by dariabezkorovaina** [comment]
Hi @damccorm, this is the final PR in these Duet AI prompts series: we've created the "no links" versions of all the "documentation lookup" prompts, as well as added some nits and polishing edits. Please take a look :)

**Review by andreydevyatkin — APPROVED** [review]
LGTM, thanks!

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**